File: advent-10/advent-10.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

void part1(vector<int> adaptors) {
    adaptors.push_back(0);
    sort(adaptors.begin(), adaptors.end());
    vector<int> diffs(4, 0);
    for (int i = 1; i < adaptors.size(); i++) {
        diffs[adaptors[i] - adaptors[i-1]]++; 
    }
    cout << "Differences Multiplied: " << diffs[1] * (diffs[3] + 1) << endl;
}

void part2(vector<int> adaptors) {
    adaptors.push_back(0);
    sort(adaptors.begin(), adaptors.end());
    unordered_map<int, long long> map;
    map[adaptors[adaptors.size()-1]] = 1;
    for (int i = adaptors.size()-2; i >= 0; i--) {
        for (int j = 1; j <= 3; j++) {
            map[adaptors[i]] += map[adaptors[i]+j];
        }
    }
    cout << "Ways: " << map[adaptors[0]] << endl;
}

int main() {
    vector<int> adaptors;
    int x;
    while (cin >> x) {
        adaptors.push_back(x);
    }
    part1(adaptors);
    part2(adaptors);
    return 0;
}

File: advent-10/advent-10.py
from collections import defaultdict

adaptors = []
with open("input.txt", 'r') as file:
    adaptors = [int(line) for line in file.readlines()]

def part1(adaptors):
    adaptors.append(0)
    adaptors = sorted(adaptors)
    diffs = [0] * 4
    for i in range(1, len(adaptors)):
        diffs[adaptors[i] - adaptors[i-1]] += 1
    print("Multiplied counts: ", diffs[1] * (diffs[3]+1))

def part2(adaptors):
    adaptors.append(0)
    adaptors = sorted(adaptors)
    d = defaultdict(int)
    d[adaptors[-1]] = 1
    for adaptor in adaptors[::-1]:
        for i in range(1,4):
            d[adaptor] += d[adaptor + i]
    print("Total Num Ways: ", d[0])

if __name__ == "__main__":
    part1(adaptors.copy())
    part2(adaptors.copy())
